{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Treehouse analysis stories coded up against the GA4GH Variant API\n",
    "\n",
    "References:\n",
    "\n",
    "https://ga4gh-schemas.readthedocs.io/en/latest/schemas/index.html\n",
    "\n",
    "https://github.com/BD2KGenomics/bioapi-examples\n",
    "\n",
    "https://github.com/ga4gh/ga4gh-client/blob/master/ga4gh/client/client.py"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ga4gh.client version 0.0.4.dev9+ngd79dd00.d20170106\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "\n",
    "!pip2 install -q --upgrade ga4gh --pre\n",
    "import ga4gh\n",
    "from ga4gh.client import client\n",
    "print \"ga4gh.client version\", ga4gh.client.__version__"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Data sets:\n",
      "[id: \"WyIxa2dlbm9tZXMiXQ\"\n",
      "name: \"1kgenomes\"\n",
      "description: \"Variants from the 1000 Genomes project and GENCODE genes annotations\"\n",
      "]\n",
      "Variant sets:\n",
      "[u'phase3-release', u'functional-annotation']\n",
      "References:\n",
      "[id: \"WyJOQ0JJMzciXQ\"\n",
      "name: \"NCBI37\"\n",
      "md5checksum: \"54e0bb53844059bb7152618fc927cfa9\"\n",
      "ncbi_taxon_id: 9606\n",
      "description: \"NCBI37 assembly of the human genome\"\n",
      "assembly_id: \"hg37\"\n",
      "source_uri: \"ftp://ftp.1000genomes.ebi.ac.uk/vol1/ftp/technical/reference/phase2_reference_assembly_sequence/hs37d5.fa.gz\"\n",
      "]\n"
     ]
    }
   ],
   "source": [
    "# Connect to the GA4GH server\n",
    "c = client.HttpClient(\"http://1kgenomes.ga4gh.org\")\n",
    "\n",
    "# Print references in this server\n",
    "print \"References:\\n\", list(c.search_reference_sets())\n",
    "\n",
    "# Use the first data set\n",
    "datasets = list(c.search_datasets())\n",
    "print \"Data sets:\\n\", datasets\n",
    "dataset_id = datasets[0].id\n",
    "\n",
    "# Use the first variant set\n",
    "variant_sets = list(c.search_variant_sets(dataset_id=dataset_id))\n",
    "print \"Variant sets:\\n\", [v.name for v in variant_sets]\n",
    "variant_set_id=variant_sets[0].id"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Call sets found: 1\n"
     ]
    }
   ],
   "source": [
    "donor_id = \"HG00356\"\n",
    "\n",
    "# call_sets = list(c.search_call_sets(variant_set_id=variant_sets[0].id, biosample_id=biosamples[0].id))\n",
    "call_sets = list(c.search_call_sets(variant_set_id=variant_set_id, name=donor_id))\n",
    "print \"Call sets found:\", len(call_sets)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Variant Stories\n",
    "As part of the overview, the analyst wants to review any of these variants present in the patient that are on this list: myCancerGenome.bed here: https://www.synapse.org/#!Synapse:syn6041633/wiki/407318. \n",
    "\n",
    "As part of the overview, the analyst wants to review  any high impact variants  present in the patient in cancer genes defined here: the Census_allWed Oct 5 22-03-15 2016.tsv list at those genes\n",
    "\n",
    "We want to add the list of genes from these two searches to our enrichment analysis.\n",
    "\n",
    "The variant coordinates included in the variant list are from GRCh37 and the gene coordinates in the gene list are from hg38. The variant input we want to match this info to is sometimes from bcbio analysis of exome or wgs, which we use grch37 for, and sometimes from variant calling on RNA-Seq, which uses hg38. The important aspect is the amino acid change, e.g. NRAS (Q61R), but it's easy got get it wrong when triangulating from different coordinates. In other words, the automation has to be validated variant by variant, and data source by data source.\n",
    "\n",
    "\n",
    "https://www.synapse.org/#!Synapse:syn6041633/wiki/407318\n",
    "\n",
    "Canonical variants to check for in every patient (\"keep\" list)\n",
    "Obtained from Max H. on October 14, 2016\n",
    "myCancerGenome.bed - coordinates from GRCh37\n",
    "\n",
    "List of cancer genes to prioritize for variant review\n",
    "Downloaded from Cancer Gene Census https://cancer.sanger.ac.uk/census by Olena on October 5, 2016:\n",
    "Census_allWed Oct 5 22-03-15 2016.tsv -- coordinates from hg38"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>stop</th>\n",
       "      <th>name</th>\n",
       "      <th>details</th>\n",
       "      <th>disease</th>\n",
       "      <th>reference</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>110881944</td>\n",
       "      <td>110881945</td>\n",
       "      <td>RBM15-MKL1 Fusion</td>\n",
       "      <td>RBM15-MKL1 Fusion</td>\n",
       "      <td>Acute Myeloid Leukemia</td>\n",
       "      <td>https://mycancergenome.org/content/disease/acu...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>115247084</td>\n",
       "      <td>115247085</td>\n",
       "      <td>NRAS No Mutation Detected</td>\n",
       "      <td>NRAS No Mutation Detected</td>\n",
       "      <td>Colorectal Cancer</td>\n",
       "      <td>https://mycancergenome.org/content/disease/col...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>115256528</td>\n",
       "      <td>115256529</td>\n",
       "      <td>Q61L Melanoma</td>\n",
       "      <td>NRAS c.182_183delAAinsTG (Q61L)</td>\n",
       "      <td>Melanoma</td>\n",
       "      <td>https://mycancergenome.org/content/disease/mel...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>115256528</td>\n",
       "      <td>115256529</td>\n",
       "      <td>Q61R Melanoma</td>\n",
       "      <td>NRAS c.182_183delAAinsGG (Q61R)</td>\n",
       "      <td>Melanoma</td>\n",
       "      <td>https://mycancergenome.org/content/disease/mel...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>115256528</td>\n",
       "      <td>115256529</td>\n",
       "      <td>Q61H Melanoma</td>\n",
       "      <td>NRAS c.183A&gt;T (Q61H)</td>\n",
       "      <td>Melanoma</td>\n",
       "      <td>https://mycancergenome.org/content/disease/mel...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  chrom      start       stop                       name  \\\n",
       "0  chr1  110881944  110881945          RBM15-MKL1 Fusion   \n",
       "1  chr1  115247084  115247085  NRAS No Mutation Detected   \n",
       "2  chr1  115256528  115256529              Q61L Melanoma   \n",
       "3  chr1  115256528  115256529              Q61R Melanoma   \n",
       "4  chr1  115256528  115256529              Q61H Melanoma   \n",
       "\n",
       "                           details                 disease  \\\n",
       "0                RBM15-MKL1 Fusion  Acute Myeloid Leukemia   \n",
       "1        NRAS No Mutation Detected       Colorectal Cancer   \n",
       "2  NRAS c.182_183delAAinsTG (Q61L)                Melanoma   \n",
       "3  NRAS c.182_183delAAinsGG (Q61R)                Melanoma   \n",
       "4             NRAS c.183A>T (Q61H)                Melanoma   \n",
       "\n",
       "                                           reference  \n",
       "0  https://mycancergenome.org/content/disease/acu...  \n",
       "1  https://mycancergenome.org/content/disease/col...  \n",
       "2  https://mycancergenome.org/content/disease/mel...  \n",
       "3  https://mycancergenome.org/content/disease/mel...  \n",
       "4  https://mycancergenome.org/content/disease/mel...  "
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Load high impact variants list (GRCh37)\n",
    "high_impact_variants = pd.read_csv(\"treehouse/myCancerGenome.bed\", sep=\"\\t\",\n",
    "                                  names=[\"chrom\", \"start\", \"stop\", \"name\", \"details\", \"disease\", \"reference\"])\n",
    "high_impact_variants.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "High impact calls found: 1\n",
      "CPU times: user 1.06 s, sys: 56 ms, total: 1.12 s\n",
      "Wall time: 4.78 s\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "# See if any of the high impact variants are present in the first call set\n",
    "high_impact_calls = [(call, hv) for i, hv in high_impact_variants.iterrows() \n",
    "                     for call in c.search_variants(variant_set_id=variant_sets[0].id, call_set_ids=[call_sets[0].id],\n",
    "                                                   reference_name=hv.chrom[3:], start=hv.start, end=hv.stop)\n",
    "                     if call.start >= hv.start and call.end <= hv.stop]\n",
    "print \"High impact calls found:\", len(high_impact_calls)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "As part of the overview, the analyst wants to review any high impact variants present in the patient in cancer genes defined here: the Census_allWed Oct 5 22-03-15 2016.tsv list at those genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene Symbol</th>\n",
       "      <th>Name</th>\n",
       "      <th>Entrez GeneId</th>\n",
       "      <th>Genome Location</th>\n",
       "      <th>Chr Band</th>\n",
       "      <th>Somatic</th>\n",
       "      <th>Germline</th>\n",
       "      <th>Tumour Types(Somatic)</th>\n",
       "      <th>Tumour Types(Germline)</th>\n",
       "      <th>Cancer Syndrome</th>\n",
       "      <th>Tissue Type</th>\n",
       "      <th>Molecular Genetics</th>\n",
       "      <th>Role in Cancer</th>\n",
       "      <th>Mutation Types</th>\n",
       "      <th>Translocation Partner</th>\n",
       "      <th>Other Germline Mut</th>\n",
       "      <th>Other Syndrome</th>\n",
       "      <th>Synonyms</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ABI1</td>\n",
       "      <td>abl-interactor 1</td>\n",
       "      <td>10006</td>\n",
       "      <td>10:26748570-26860863</td>\n",
       "      <td>10p11.2</td>\n",
       "      <td>yes</td>\n",
       "      <td>NaN</td>\n",
       "      <td>AML</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>L</td>\n",
       "      <td>Dom</td>\n",
       "      <td>TSG</td>\n",
       "      <td>T</td>\n",
       "      <td>KMT2A</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ABI1,E3B1,ABI-1,SSH3BP1,10006</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ABL1</td>\n",
       "      <td>v-abl Abelson murine leukemia viral oncogene h...</td>\n",
       "      <td>25</td>\n",
       "      <td>9:130835447-130885683</td>\n",
       "      <td>9q34.1</td>\n",
       "      <td>yes</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CML, ALL, T-ALL</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>L</td>\n",
       "      <td>Dom</td>\n",
       "      <td>oncogene</td>\n",
       "      <td>T, Mis</td>\n",
       "      <td>BCR, ETV6, NUP214</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ABL1,p150,ABL,c-ABL,JTK7,bcr/abl,v-abl,P00519,...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ABL2</td>\n",
       "      <td>c-abl oncogene 2, non-receptor tyrosine kinase</td>\n",
       "      <td>27</td>\n",
       "      <td>1:179107718-179143044</td>\n",
       "      <td>1q24-q25</td>\n",
       "      <td>yes</td>\n",
       "      <td>NaN</td>\n",
       "      <td>AML</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>L</td>\n",
       "      <td>Dom</td>\n",
       "      <td>oncogene</td>\n",
       "      <td>T</td>\n",
       "      <td>ETV6</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ABL2,ARG,RP11-177A2_3,ABLL,P42684,ENSG00000143...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ACKR3</td>\n",
       "      <td>atypical chemokine receptor 3</td>\n",
       "      <td>57007</td>\n",
       "      <td>2:-</td>\n",
       "      <td>2q37.3</td>\n",
       "      <td>yes</td>\n",
       "      <td>NaN</td>\n",
       "      <td>lipoma</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>M</td>\n",
       "      <td>Dom</td>\n",
       "      <td>oncogene</td>\n",
       "      <td>T</td>\n",
       "      <td>HMGA2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ACSL3</td>\n",
       "      <td>acyl-CoA synthetase long-chain family member 3</td>\n",
       "      <td>2181</td>\n",
       "      <td>2:222908773-222941654</td>\n",
       "      <td>2q36</td>\n",
       "      <td>yes</td>\n",
       "      <td>NaN</td>\n",
       "      <td>prostate</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>E</td>\n",
       "      <td>Dom</td>\n",
       "      <td>NaN</td>\n",
       "      <td>T</td>\n",
       "      <td>ETV1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2181,PRO2194,ACS3,FACL3,O95573,ENSG00000123983...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  Gene Symbol                                               Name  \\\n",
       "0        ABI1                                   abl-interactor 1   \n",
       "1        ABL1  v-abl Abelson murine leukemia viral oncogene h...   \n",
       "2        ABL2     c-abl oncogene 2, non-receptor tyrosine kinase   \n",
       "3       ACKR3                      atypical chemokine receptor 3   \n",
       "4       ACSL3     acyl-CoA synthetase long-chain family member 3   \n",
       "\n",
       "   Entrez GeneId        Genome Location  Chr Band Somatic Germline  \\\n",
       "0          10006   10:26748570-26860863   10p11.2     yes      NaN   \n",
       "1             25  9:130835447-130885683    9q34.1     yes      NaN   \n",
       "2             27  1:179107718-179143044  1q24-q25     yes      NaN   \n",
       "3          57007                    2:-    2q37.3     yes      NaN   \n",
       "4           2181  2:222908773-222941654      2q36     yes      NaN   \n",
       "\n",
       "  Tumour Types(Somatic) Tumour Types(Germline) Cancer Syndrome Tissue Type  \\\n",
       "0                   AML                    NaN             NaN           L   \n",
       "1       CML, ALL, T-ALL                    NaN             NaN           L   \n",
       "2                   AML                    NaN             NaN           L   \n",
       "3                lipoma                    NaN             NaN           M   \n",
       "4              prostate                    NaN             NaN           E   \n",
       "\n",
       "  Molecular Genetics Role in Cancer Mutation Types Translocation Partner  \\\n",
       "0                Dom            TSG              T                 KMT2A   \n",
       "1                Dom       oncogene         T, Mis     BCR, ETV6, NUP214   \n",
       "2                Dom       oncogene              T                  ETV6   \n",
       "3                Dom       oncogene              T                 HMGA2   \n",
       "4                Dom            NaN              T                  ETV1   \n",
       "\n",
       "  Other Germline Mut Other Syndrome  \\\n",
       "0                NaN            NaN   \n",
       "1                NaN            NaN   \n",
       "2                NaN            NaN   \n",
       "3                NaN            NaN   \n",
       "4                NaN            NaN   \n",
       "\n",
       "                                            Synonyms  \n",
       "0                      ABI1,E3B1,ABI-1,SSH3BP1,10006  \n",
       "1  ABL1,p150,ABL,c-ABL,JTK7,bcr/abl,v-abl,P00519,...  \n",
       "2  ABL2,ARG,RP11-177A2_3,ABLL,P42684,ENSG00000143...  \n",
       "3                                                NaN  \n",
       "4  2181,PRO2194,ACS3,FACL3,O95573,ENSG00000123983...  "
      ]
     },
     "execution_count": 47,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Cancer genes\n",
    "cancer_genes = pd.read_csv(\"treehouse/Census_allWed Oct  5 22-03-15 2016.tsv\", sep=\"\\t\")\n",
    "cancer_genes.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "metadata": {},
   "outputs": [
    {
     "ename": "AttributeError",
     "evalue": "'NoneType' object has no attribute 'groups'",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m\u001b[0m",
      "\u001b[1;31mAttributeError\u001b[0mTraceback (most recent call last)",
      "\u001b[1;32m<ipython-input-57-95411b3cf691>\u001b[0m in \u001b[0;36m<module>\u001b[1;34m()\u001b[0m\n\u001b[1;32m----> 1\u001b[1;33m \u001b[0mget_ipython\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mrun_cell_magic\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;34mu'time'\u001b[0m\u001b[1;33m,\u001b[0m \u001b[1;34mu''\u001b[0m\u001b[1;33m,\u001b[0m \u001b[1;34mu'def calls_in_cancer_genes(cancer_genes, server, variant_set_id, call_set_id):\\n    for _, gene in cancer_genes.iterrows():\\n        reference_name, start, stop = re.match(\"(\\\\d+)\\\\:(\\\\d+)-(\\\\d+)\", gene[\"Genome Location\"]).groups()\\n        start = int(start)\\n        stop = int(stop)\\n        for call in server.search_variants(variant_set_id=variant_set_id, call_set_ids=[call_set_id],\\n                                           reference_name=reference_name, start=start, end=stop):\\n            if call.start >= start and call.end <= hv.stop:\\n                yield call\\n                \\nlist(calls_in_cancer_genes(cancer_genes, c, variant_sets[0].id, call_sets[0].id))'\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[1;32m/opt/conda/envs/python2/lib/python2.7/site-packages/IPython/core/interactiveshell.pyc\u001b[0m in \u001b[0;36mrun_cell_magic\u001b[1;34m(self, magic_name, line, cell)\u001b[0m\n\u001b[0;32m   2118\u001b[0m             \u001b[0mmagic_arg_s\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mself\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mvar_expand\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mline\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mstack_depth\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m   2119\u001b[0m             \u001b[1;32mwith\u001b[0m \u001b[0mself\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mbuiltin_trap\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m-> 2120\u001b[1;33m                 \u001b[0mresult\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mfn\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mmagic_arg_s\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mcell\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m   2121\u001b[0m             \u001b[1;32mreturn\u001b[0m \u001b[0mresult\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m   2122\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;32m<decorator-gen-60>\u001b[0m in \u001b[0;36mtime\u001b[1;34m(self, line, cell, local_ns)\u001b[0m\n",
      "\u001b[1;32m/opt/conda/envs/python2/lib/python2.7/site-packages/IPython/core/magic.pyc\u001b[0m in \u001b[0;36m<lambda>\u001b[1;34m(f, *a, **k)\u001b[0m\n\u001b[0;32m    191\u001b[0m     \u001b[1;31m# but it's overkill for just that one bit of state.\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m    192\u001b[0m     \u001b[1;32mdef\u001b[0m \u001b[0mmagic_deco\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0marg\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m--> 193\u001b[1;33m         \u001b[0mcall\u001b[0m \u001b[1;33m=\u001b[0m \u001b[1;32mlambda\u001b[0m \u001b[0mf\u001b[0m\u001b[1;33m,\u001b[0m \u001b[1;33m*\u001b[0m\u001b[0ma\u001b[0m\u001b[1;33m,\u001b[0m \u001b[1;33m**\u001b[0m\u001b[0mk\u001b[0m\u001b[1;33m:\u001b[0m \u001b[0mf\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;33m*\u001b[0m\u001b[0ma\u001b[0m\u001b[1;33m,\u001b[0m \u001b[1;33m**\u001b[0m\u001b[0mk\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m    194\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m    195\u001b[0m         \u001b[1;32mif\u001b[0m \u001b[0mcallable\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0marg\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;32m/opt/conda/envs/python2/lib/python2.7/site-packages/IPython/core/magics/execution.pyc\u001b[0m in \u001b[0;36mtime\u001b[1;34m(self, line, cell, local_ns)\u001b[0m\n\u001b[0;32m   1175\u001b[0m         \u001b[1;32melse\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m   1176\u001b[0m             \u001b[0mst\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mclock2\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m-> 1177\u001b[1;33m             \u001b[1;32mexec\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mcode\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mglob\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mlocal_ns\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m   1178\u001b[0m             \u001b[0mend\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mclock2\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m   1179\u001b[0m             \u001b[0mout\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mNone\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;32m<timed exec>\u001b[0m in \u001b[0;36m<module>\u001b[1;34m()\u001b[0m\n",
      "\u001b[1;32m<timed exec>\u001b[0m in \u001b[0;36mcalls_in_cancer_genes\u001b[1;34m(cancer_genes, server, variant_set_id, call_set_id)\u001b[0m\n",
      "\u001b[1;31mAttributeError\u001b[0m: 'NoneType' object has no attribute 'groups'"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "def calls_in_cancer_genes(cancer_genes, server, variant_set_id, call_set_id):\n",
    "    for _, gene in cancer_genes.iterrows():\n",
    "        reference_name, start, stop = re.match(\"(\\d+)\\:(\\d+)-(\\d+)\", gene[\"Genome Location\"]).groups()\n",
    "        start = int(start)\n",
    "        stop = int(stop)\n",
    "        for call in server.search_variants(variant_set_id=variant_set_id, call_set_ids=[call_set_id],\n",
    "                                           reference_name=reference_name, start=start, end=stop):\n",
    "            if call.start >= start and call.end <= hv.stop:\n",
    "                yield call\n",
    "                \n",
    "list(calls_in_cancer_genes(cancer_genes, c, variant_sets[0].id, call_sets[0].id))"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 1
}
